File: backend/app/schemas/sso.py
import uuid
from typing import Optional
from pydantic import BaseModel, ConfigDict, EmailStr

from app.schemas.user import UserRead


class JanParichayClaims(BaseModel):
    """
    Standard claims payload returned by NIC MeriPehchan / Jan Parichay OIDC service.
    """
    parichay_id: str
    user_id: Optional[str] = None
    full_name: str
    email: str
    department: str
    designation: str
    state_code: Optional[str] = None
    service_id: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class SSOSandboxProfile(BaseModel):
    """
    Pre-configured mock government officer persona available in sandbox mode.
    """
    id: str
    full_name: str
    email: str
    designation: str
    department: str
    state_code: str
    mapped_app_role: str
    description: str


class SSOInitResponse(BaseModel):
    """
    Response returned to frontend to initiate government OIDC login flow.
    """
    authorization_url: str
    state: str
    code_verifier: Optional[str] = None
    code_challenge: Optional[str] = None  # Deprecated alias for backward compatibility
    is_sandbox: bool


class SSOCallbackRequest(BaseModel):
    """
    Authorization code payload posted by frontend upon return from MeriPehchan.
    """
    code: str
    state: str
    code_verifier: Optional[str] = None


class SSOSandboxAuthorizeRequest(BaseModel):
    """
    Request to approve authentication as a specific sandbox persona.
    """
    persona_id: str
    state: str


class SSOStatusResponse(BaseModel):
    """
    Health and configuration status of the government SSO gateway.
    """
    enabled: bool
    mode: str  # 'live' or 'sandbox'
    provider_name: str = "MeriPehchan (Jan Parichay) NSSO"
    discovery_url: str
    client_id_configured: bool


class SSOAuthResponse(BaseModel):
    """
    Authentication response with JWT tokens and provisioned officer identity.
    """
    access_token: str
    refresh_token: str
    token_type: str = "bearer"
    user: UserRead
    claims: Optional[JanParichayClaims] = None

File: frontend/app/types/sso.ts
export interface JanParichayClaims {
  parichay_id: string;
  user_id?: string | null;
  full_name: string;
  email: string;
  department: string;
  designation: string;
  state_code?: string | null;
  service_id?: string | null;
}

export interface SSOSandboxProfile {
  id: string;
  full_name: string;
  email: string;
  designation: string;
  department: string;
  state_code: string;
  mapped_app_role: 'officer' | 'supervisor' | 'admin';
  description: string;
}

export interface SSOInitResponse {
  authorization_url: string;
  state: string;
  code_verifier?: string | null;
  code_challenge?: string | null;
  is_sandbox: boolean;
}

export interface SSOStatusResponse {
  enabled: boolean;
  mode: 'live' | 'sandbox';
  provider_name: string;
  discovery_url: string;
  client_id_configured: boolean;
}

export interface SSOAuthResponse {
  access_token: string;
  refresh_token: string;
  token_type: string;
  user: {
    id: string;
    email: string;
    full_name: string;
    role: string;
    region?: string | null;
    is_active: boolean;
  };
  claims?: JanParichayClaims | null;
}
